p2p/host/basic/natmgr.go

```diff
@@ -87,151 +90,151 @@ func (nmgr *natManager) discoverNAT() {
+		if err != nil {
+			log.Error("DiscoverNAT error:", err)
+			close(nmgr.ready)
 			return
-		case <-discoverdone:
-			if nat == nil { // no nat, or failed to get it.
-				return
-			}
 		}
 
-		// wire up the nat to close when nmgr closes.
-		// nmgr.proc is our parent, and waiting for us.
-		nmgr.proc.AddChild(nat.Process())
-
-		// set the nat.
 		nmgr.natmu.Lock()
-		nmgr.nat = nat
+		nmgr.nat = natInstance
 		nmgr.natmu.Unlock()
-
-		// signal that we're ready to process nat mappings:
 		close(nmgr.ready)
 
+		// wire up the nat to close when nmgr closes.
+		// nmgr.proc is our parent, and waiting for us.
+		nmgr.proc.AddChild(nmgr.nat.Process())
+
 		// sign natManager up for network notifications
 		// we need to sign up here to avoid missing some notifs
 		// before the NAT has been found.
 		nmgr.net.Notify((*nmgrNetNotifiee)(nmgr))
-
-		// if any interfaces were brought up while we were setting up
-		// the nat, now is the time to setup port mappings for them.
-		// we release ready, then grab them to avoid losing any. adding
-		// a port mapping is idempotent, so its ok to add the same twice.
-		addrs := nmgr.net.ListenAddresses()
-		for _, addr := range addrs {
-			// we do it async because it's slow and we may want to close beforehand
-			go addPortMapping(nmgr, addr)
-		}
+		nmgr.sync()
 	})
 }
 
-// NAT returns the natManager's nat object. this may be nil, if
-// (a) the search process is still ongoing, or (b) the search process
-// found no nat. Clients must check whether the return value is nil.
-func (nmgr *natManager) NAT() *inat.NAT {
-	nmgr.natmu.Lock()
-	defer nmgr.natmu.Unlock()
-	return nmgr.nat
-}
-
-func addPortMapping(nmgr *natManager, intaddr ma.Multiaddr) {
+// syncs the current NAT mappings, removing any outdated mappings and adding any
+// new mappings.
+func (nmgr *natManager) sync() {
 	nat := nmgr.NAT()
 	if nat == nil {
-		panic("natManager addPortMapping called without a nat.")
+		// Nothing to do.
+		return
 	}
 
-	// first, check if the port mapping already exists.
-	for _, mapping := range nat.Mappings() {
-		if mapping.InternalAddr().Equal(intaddr) {
-			return // it exists! return.
+	nmgr.proc.Go(func(_ goprocess.Process) {
+		nmgr.syncMu.Lock()
+		defer nmgr.syncMu.Unlock()
+
+		ports := map[string]map[int]bool{
+			"tcp": map[int]bool{},
+			"udp": map[int]bool{},
 		}
-	}
+		for _, maddr := range nmgr.net.ListenAddresses() {
+			// Strip the IP
+			maIP, rest := ma.SplitFirst(maddr)
+			if maIP == nil || rest == nil {
+				continue
+			}
 
-	ctx := context.TODO()
-	lm := make(lgbl.DeferredMap)
-	lm["internalAddr"] = func() interface{} { return intaddr.String() }
+			switch maIP.Protocol().Code {
+			case ma.P_IP6, ma.P_IP4:
+			default:
+				continue
+			}
 
-	defer log.EventBegin(ctx, "natMgrAddPortMappingWait", lm).Done()
+			// Only bother if we're listening on a
+			// unicast/unspecified IP.
+			ip := net.IP(maIP.RawValue())
+			if !(ip.IsGlobalUnicast() || ip.IsUnspecified()) {
```

Comment:
Just to confirm. This doesn't `continue` if the listen address is public (i.e. we don't need a port mapping). That's because we're presuming in that case there's no NAT device at all, and hence this func would return at line 124? What would happen if the host is listening on multiple network interfaces, of which one is public and one is private (behind a NAT)?